core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -143,8 +143,18 @@ private[spark] class TaskSetManager(
   private var calculatedTasks = 0
 
   private[scheduler] val taskSetExcludelistHelperOpt: Option[TaskSetExcludelist] = {
-    healthTracker.map { _ =>
-      new TaskSetExcludelist(sched.sc.listenerBus, conf, stageId, taskSet.stageAttemptId, clock)
+    if (TaskSetExcludelist.isExcludeOnFailureEnabled(conf)) {
+      Some(new TaskSetExcludelist(sched.sc.listenerBus, conf, stageId,
+        taskSet.stageAttemptId, clock))
+    } else if (healthTracker.isDefined) {
+      // If we enabled exclusion at application level but not at taskset level exclusion, we create
+      // TaskSetExcludelist in dry run mode.
```

Comment:
This seems to introduce the behaviour change. Before this PR,  `TaskSetExcludelist` is created in non-dry-run mode when `EXCLUDE_ON_FAILURE_ENABLED` is on. After this PR, the non-dry-run mode creation requires another conf `EXCLUDE_ON_FAILURE_ENABLED_TASK_AND_STAGE` on implicitly. I think adding an explicit conf for "_whether creating TaskSetExcludelist in dry-run mode_" maybe a better solution.

Reply:
This would not introduce a behavior change.
`EXCLUDE_ON_FAILURE_ENABLED_TASK_AND_STAGE` is an optional config and would default back to current behavior if not set. So for existing users they can be unaware of this change, and setting `EXCLUDE_ON_FAILURE_ENABLED` would bring both `HealthTracker` and `TaskSetExcludeList` on.

Reply:
SGTM